src/Symfony/Component/HttpClient/Response/AmpResponse.php

```diff
@@ -0,0 +1,326 @@
+            $activity[$id][] = $e;
+        }
+
+        Loop::stop();
+    }
+
+    private static function followRedirects(Response $response, array &$info, array &$headers, HttpClient $client, int $maxRedirects, CancellationTokenSource $canceller, Request $originRequest)
+    {
+        $previousUrl = null;
+
+        while (true) {
+            $info['url'] = (string) $response->getRequest()->getUri();
+            $info['redirect_url'] = null;
+            $status = $response->getStatus();
+
+            $responseHeaders = [
+                sprintf('HTTP/%s %s %s', $response->getProtocolVersion(), $status, $response->getReason()),
+            ];
+
+            foreach ($response->getHeaders() as $name => $values) {
+                foreach ($values as $v) {
+                    $responseHeaders[] = ucwords($name, '-').': '.$v;
+                }
+            }
+
+            self::addResponseHeaders($responseHeaders, $info, $headers, $info['debug']);
+
+            if (!\in_array($status, [301, 302, 303, 307, 308], true) || null === $location = $response->getHeader('location')) {
+                return $response;
+            }
+
+            $urlResolver = new class() {
+                use HttpClientTrait {
+                    parseUrl as public;
+                    resolveUrl as public;
+                }
+            };
+
+            try {
+                $previousUrl = $previousUrl ?? $urlResolver::parseUrl($info['url']);
+                $location = $urlResolver::parseUrl($location);
+                $location = $urlResolver::resolveUrl($location, $previousUrl);
+                $info['redirect_url'] = implode('', $location);
+            } catch (InvalidArgumentException $e) {
+                return $response;
+            }
+
+            if (0 >= $maxRedirects || $info['redirect_count'] >= $maxRedirects) {
+                return $response;
+            }
+
+            try {
+                // Discard body of redirects
+                while (null !== yield $response->getBody()->read()) {
+                }
+            } catch (HttpException | StreamException $e) {
+                // Ignore streaming errors on previous responses
+            }
+
+            ++$info['redirect_count'];
+            $previousUrl = $location;
+            $request = new Request($info['redirect_url'], $response->getRequest()->getMethod());
+            $request->setProtocolVersions($originRequest->getProtocolVersions());
+            $request->setTcpConnectTimeout($originRequest->getTcpConnectTimeout());
+            $request->setTlsHandshakeTimeout($originRequest->getTlsHandshakeTimeout());
+            $request->setTransferTimeout($originRequest->getTransferTimeout());
+
+            if (\in_array($status, [301, 302, 303], true)) {
+                $originRequest->removeHeader('transfer-encoding');
+                $originRequest->removeHeader('content-length');
+                $originRequest->removeHeader('content-type');
```

Comment:
Why do you modify the original request here?

Reply:
Because this state change needs to be kept when several redirections happen, and this does it.
Note that the $originRequest object is internal so this has no side-effects.